File: Can you sort the array - GFG/can-you-sort-the-array.java
//{ Driver Code Starts
// Initial Template for Java
import java.io.*;
import java.util.*;


// } Driver Code Ends
// User function Template for Java
class Solution {
    public boolean canSort(int[] arr, int[] frozen) {
        int countZeroes = 0;
        
        for(int num:arr){
            if(num == 0) countZeroes++;
        }
        
        for(int i = 0;i<arr.length;i++){
            if(i<countZeroes && arr[i] == 1 && frozen[i] == 1){
                return false;
            }
            
            if(i>=countZeroes && arr[i] == 0 && frozen[i] == 1){
                return false;
            }
        }
        
        return true;
    }
}

//{ Driver Code Starts.

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(br.readLine()); // Read number of test cases
        while (t-- > 0) {
            // Read the arr array
            String[] arrInput = br.readLine().split(" ");
            int[] arr = Arrays.stream(arrInput).mapToInt(Integer::parseInt).toArray();

            // Read the locked array
            String[] frozen = br.readLine().split(" ");
            int[] locked = Arrays.stream(frozen).mapToInt(Integer::parseInt).toArray();

            Solution solution = new Solution();
            boolean result = solution.canSort(arr, locked);
            System.out.println(result ? "true" : "false");
            System.out.println("~");
        }
    }
}

// } Driver Code Ends

File: Can you sort the array - GFG/can-you-sort-the-array.cpp
//{ Driver Code Starts
// Initial Template for C++
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
// User function Template for C++
class Solution {
  public:
    bool canSort(vector<int> &arr, vector<int> &frozen) {
        int n = arr.size();
        int cntZero = count(arr.begin(),arr.end(),0);
        
        for(int i = 0;i<n;i++){
            if(i<cntZero && arr[i] == 1 && frozen[i]){
                return false;
            }
            
            if(i>=cntZero && arr[i] == 0 && frozen[i]){
                return false;
            }
        }
        return true;
    }
};


//{ Driver Code Starts.

int main() {
    string str;
    getline(cin, str);
    int t = stoi(str);
    while (t--) {
        vector<int> arr, frozen;
        int x;

        getline(cin, str);
        stringstream ss1(str);
        while (ss1 >> x)
            arr.push_back(x);

        getline(cin, str);
        stringstream ss2(str);
        while (ss2 >> x)
            frozen.push_back(x);

        Solution ob;
        bool ans = ob.canSort(arr, frozen);
        if (ans) {
            cout << "true" << endl;
        } else {
            cout << "false" << endl;
        }
        cout << "~" << endl;
    }

    return 0;
}

// } Driver Code Ends